tests/integration/conftest.py

```diff
@@ -142,40 +79,90 @@ def fixture_kafka_server(request, session_tmppath: Path) -> Iterator[KafkaServer
+                maybe_download_kafka(kafka_description)
+
+                zk_client_port = get_random_port(port_range=ZK_PORT_RANGE, blacklist=[])
+                zk_admin_port = get_random_port(port_range=ZK_PORT_RANGE, blacklist=[zk_client_port])
+                zk_config = ZKConfig(
+                    client_port=zk_client_port,
+                    admin_port=zk_admin_port,
+                    path=str(zk_dir),
+                )
+
+                zk_proc = configure_and_start_zk(zk_config, kafka_description)
                 stack.callback(stop_process, zk_proc)
 
                 # Make sure zookeeper is running before trying to start Kafka
-                wait_for_port(zk_config.client_port, wait_time=20)
-
-                kafka_config, kafka_proc = configure_and_start_kafka(kafka_dir, zk_config)
+                wait_for_port_subprocess(zk_config.client_port, zk_proc, wait_time=20)
+
+                kafka_plaintext_port = get_random_port(port_range=KAFKA_PORT_RANGE, blacklist=[])
+                data_dir = session_datadir / "kafka"
+                log_dir = session_logdir / "kafka"
+                data_dir.mkdir(parents=True)
+                log_dir.mkdir(parents=True)
+                kafka_config = KafkaConfig(
+                    datadir=str(data_dir),
+                    logdir=str(log_dir),
+                    plaintext_port=kafka_plaintext_port,
+                )
+                kafka_proc = configure_and_start_kafka(
+                    zk_config=zk_config,
+                    kafka_config=kafka_config,
+                    kafka_description=kafka_description,
+                    log4j_config=KAFKA_LOG4J,
+                )
                 stack.callback(stop_process, kafka_proc)
 
                 config_data = {
                     "zookeeper": asdict(zk_config),
                     "kafka": asdict(kafka_config),
+                    WORKER_COUNTER_KEY: 1,
                 }
                 transfer_file.write_text(ujson.dumps(config_data))
 
         # Make sure every test worker can communicate with kafka
-        kafka_servers = KafkaServers(bootstrap_servers=[f"127.0.0.1:{kafka_config.kafka_port}"])
+        kafka_servers = KafkaServers(bootstrap_servers=[f"127.0.0.1:{kafka_config.plaintext_port}"])
         wait_for_kafka(kafka_servers, KAFKA_WAIT_TIMEOUT)
+
         yield kafka_servers
+
+        # Signal the worker finished
+        with FileLock(str(lock_file)):  # pylint: disable=abstract-class-instantiated
+            assert transfer_file.exists(), "transfer_file disappeared"
+            config_data = ujson.loads(transfer_file.read_text())
+            config_data[WORKER_COUNTER_KEY] -= 1
+            transfer_file.write_text(ujson.dumps(config_data))
+
+        # Wait until every worker finished before stopping the servers
+        worker_counter = float("inf")
+        while worker_counter > 0:
+            with FileLock(str(lock_file)):  # pylint: disable=abstract-class-instantiated
```

Comment:
This was a bug. The "leader" node could finish running the tests before the other workers, and it would stop the server before the tests finished running.

I verified this in a very rudimentary way:

- Made sure gw1 was the leader by adding a sleep of 0.5s for gw0 before acquiring the lock
- Picked the test_union_to_union and made gw0 wait for 30s before running the test, this makes sure gw1 finish running the tests and stops kafka before gw0 starts
- The above above gives a connection error when gw0 runs the test